pkg/containerwatcher/v1/container_watcher_private.go

```diff
@@ -69,7 +76,93 @@ func (ch *IGContainerWatcher) containerCallback(notif containercollection.PubSub
 			helpers.String("k8s workload", k8sContainerID))
 		ch.preRunningContainersIDs.Remove(notif.Container.Runtime.ContainerID)
 		ch.timeBasedContainers.Remove(notif.Container.Runtime.ContainerID)
+		ch.objectCache.K8sObjectCache().DeleteSharedContainerData(notif.Container.Runtime.ContainerID)
+	}
+}
+
+func (ch *IGContainerWatcher) setSharedWatchedContainerData(container *containercollection.Container) {
+	// don't start monitoring until we have the instanceID - need to retry until the Pod is updated
+	var sharedWatchedContainerData *utils.WatchedContainerData
+	err := backoff.Retry(func() error {
+		data, err := ch.getSharedWatchedContainerData(container)
+		if err != nil {
+			return err
+		}
+		if data == nil {
+			return fmt.Errorf("received nil container data")
+		}
+		sharedWatchedContainerData = data
+		return nil
+	}, backoff.NewExponentialBackOff())
+
+	if err != nil {
+		logger.L().Error("IGContainerWatcher.containerCallback - error getting shared watched container data", helpers.Error(err))
+		return // Exit early on error
+	}
+
+	if sharedWatchedContainerData == nil {
+		logger.L().Error("IGContainerWatcher.containerCallback - shared watched container data is nil after retry")
+		return
+	}
+
+	ch.objectCache.K8sObjectCache().SetSharedContainerData(container.Runtime.ContainerID, sharedWatchedContainerData)
+}
+
+func (ch *IGContainerWatcher) getSharedWatchedContainerData(container *containercollection.Container) (*utils.WatchedContainerData, error) {
+	watchedContainer := utils.WatchedContainerData{
+		ContainerID: container.Runtime.ContainerID,
+		ImageID:     container.Runtime.ContainerImageDigest,
+		ImageTag:    container.Runtime.ContainerImageName,
+	}
+
+	wl, err := ch.k8sClient.GetWorkload(container.K8s.Namespace, "Pod", container.K8s.PodName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get workload: %w", err)
+	}
+	pod := wl.(*workloadinterface.Workload)
+	// fill container type, index and names
+	if watchedContainer.ContainerType == utils.Unknown {
+		if err := watchedContainer.SetContainerInfo(pod, container.K8s.ContainerName); err != nil {
+			return nil, fmt.Errorf("failed to set container info: %w", err)
+		}
+	}
+	// get pod template hash
+	watchedContainer.TemplateHash, _ = pod.GetPodLabel("pod-template-hash")
+	// find parentWlid
+	kind, name, err := ch.k8sClient.CalculateWorkloadParentRecursive(pod)
```

Comment:
ok so this is level 1 optimization, in a later PR we will rewrite this with calling workload.GetOwnerReferences() once and use the slice in the k8sapi shit